### PR TITLE
Support Gherkin Rule/Example syntax

### LINKED ISSUE #293: Lacks support for Gherkin 6 syntax
**Describe the bug**
When using keywords added in Gherkin 6, no special behavior occurs.

**To Reproduce**
Type any of the following keywords:
• Rule
• Example
• Scenario Template

They will not be highlighted, suggested, autocompleted, or formatted.

**Expected behavior**
Type the keywords listed above and they should be treated appropriately for their respective functions, just as other Gherkin keywords are.

### PR DESCRIPTION
**Pull request by davidjgoss**
Adds support for Rule keyword for formating,  syntax highlighting and snippets.

Fixes #293

### CODE REVIEW

File: gserver/src/format.ts [hunk 1]
```diff
@@ -67,22 +79,28 @@ export function correctIndents(text, indent, settings: Settings) {
                 }
             }
             //Now we should find current line format
-            const format = findFormat(line, settings);
+            const format: ResolvedFormat = findFormat(line, settings);
+            if (format && format.symbol === 'Rule:') {
+                currentIndentation = defaultIndentation;
```
Comment (davidjgoss) [inline gserver/src/format.ts:84]: This extra bit of logic is because the `FORMAT_CONF` is no longer enough on its own to decide how much to indent e.g. a `Given` - we need to know if we are within a `Rule` (which adds another level) or not.

File: gclient/syntaxes/feature.tmLanguage [hunk 2]
```diff
@@ -139,7 +139,7 @@
...
...|Senaryo taslağı|Senaryo|Scénář|Scénario|Schema dello scenario|Scenārijs pēc parauga|Scenārijs|Scenár|Scenariusz|Scenariul de şablon|Scenariul de sablon|Scenariu|Scenarios|Scenario Outline|Scenario Amlinellol|Scenario|Example|Scenarijus|Scenariji|Scenarijaus šablonas|Scenarijai|Scenarij|Scenarie|Rerefons|Raamstsenaarium|Příklady|Példák|Príklady|Przykłady|Primjeri|Primeri|Primer|Pozadí|Pozadina|Pozadie|Plan du scénario|Plan du Scénario|Piemēri|Pavyzdžiai|Paraugs|Osnova scénáře|Osnova|Náčrt Scénáře|Náčrt Scenáru|Mate|MISHUN SRSLY|MISHUN|Kịch bản|Kontext|Konteksts|Kontekstas|Kontekst|Koncept|Khung tình huống|Khung kịch bản|Juhtumid|Háttér|Grundlage|Geçmiş|Forgatókönyv vázlat|Forgatókönyv|Exemplos|Exemples|Exemplele|Exempel|Examples|Esquema do Cenário|Esquema do Cenario|Esquema del escenario|Esquema de l'escenari|Esempi|Escenario|Escenari|Enghreifftiau|Eksempler|Ejemplos|EXAMPLZ|Dữ liệu|Dis is what went down|Dasar|Contoh|Contexto|Contexte|Contesto|Condiţii|Conditii|Cobber|Cenário|Cenario|Cefndir|Bối cảnh|Blokes|Beispiele|Bakgrunn|Bakgrund|Baggrund|Background|B4|Antecedents|Antecedentes|All y'all|Achtergrond|Abstrakt Scenario|Abstract Scenario|Rule|Regla|Règle|Regel|Regra):(.*)</string>
```
Comment (davidjgoss) [inline gclient/syntaxes/feature.tmLanguage:142]: Changes taken from https://github.com/cucumber/cucumber/blob/master/gherkin/gherkin-languages.json

Incidentally that file is available in the published `@cucumber/gherkin` package so could potentially add a build step to generate the `feature.tmLanguage` file based on that. Happy to do another PR for that if you'd like.

**Comment by davidjgoss** [comment]
@alexkrechik did you have any feedback on this? With Rule support starting to land in Cucumber implementations would be great to have this available to users.

File: gserver/src/format.ts [hunk 3]
```diff
@@ -67,22 +79,28 @@ export function correctIndents(text, indent, settings: Settings) {
...
-                    const nextLineFormat = findFormat(nextLine, settings);
-                    indentCount = nextLineFormat === null ? defaultIndentation : nextLineFormat;
+                    const nextLineIndentation = findIndentation(nextLine, settings);
+                    indentCount = nextLineIndentation === null ? defaultIndentation : nextLineIndentation;
                 } else {
                     indentCount = defaultIndentation;
                 }
             }
+            if (format && format.symbol === 'Rule:') {
+                currentIndentation++;
```
Comment (alexkrechik) [inline gserver/src/format.ts:102]: Looks like, that indentation should be increased by 'rule' value, not by 1.

File: gserver/test/format.spec.ts [hunk 4]
```diff
@@ -15,11 +15,18 @@ const settings: any = {
...
-    let beforeUArr = beforeU.split(/\r?\n/);
-    let before = clearText(format('\t', beforeU, settings)).split(/\r?\n/);
-    it(`should not change lines num`, () => expect(before.length).to.be.equal(after.length));
-    beforeUArr.forEach((l, i) => it(`should correctly format "${l}" line`, () =>
-        expect(before[i]).to.be.equal(after[i])));
-});
\ No newline at end of file
+    [
+        {name: 'general', settings: userSettings},
+        {name: 'rule', settings: undefined}
```
Comment (alexkrechik) [inline gserver/test/format.spec.ts:20]: Why should `settings` be optional? Could we pass `userSettings` here?

File: gserver/src/format.ts (hunk 1 above)
Comment (alexkrechik) [reply gserver/src/format.ts:84]: Could we have a rule inside the rule? If not - I think, that it would be better to add some bool var like `commentsMode ` (ex. `ruleMode`) and, if it's true - add rule indentation to all the strings except `Rule` itself.

File: gclient/syntaxes/feature.tmLanguage (hunk 2 above)
Comment (alexkrechik) [reply gclient/syntaxes/feature.tmLanguage:142]: Sorry for the long response. Great idea! Maybe some command in package.json would be the best solution. But this file was edited by many contributors so we should sure, that all the words still be present here.
Also, it would be great to import all the gherkin words here - https://github.com/alexkrechik/VSCucumberAutoComplete/blob/master/gserver/src/gherkin.ts. I'll think about this.

File: gserver/src/format.ts (hunk 3 above)
Comment (davidjgoss) [reply gserver/src/format.ts:102]: Agreed, will change

File: gserver/src/format.ts (hunk 1 above)
Comment (davidjgoss) [reply gserver/src/format.ts]: Sorry for late reply - `Rule`s can't be nested no

**Comment by davidjgoss** [comment]
@alexkrechik thanks so much for finishing this up and merging

**Comment by alexkrechik** [comment]
> @alexkrechik thanks so much for finishing this up and merging

Thank you too!
